File: python-way/com/crack/snap/make/medium/stack/evaluate_reverse_polish_notation.py
# @author Mohan Sharma
from typing import List

def evalRPN(tokens: List[str]) -> int:
	stack = []
	for token in tokens:
		if token in ["+", "-", "/", "*"]:
			if len(stack) < 2:
				raise Exception("Invalid RPN expression")
			b = stack.pop()
			a = stack.pop()
			stack.append(int(eval(f"{a}{token}{b}")))
		else:
			stack.append(int(token))
	if len(stack) != 1:
		raise Exception("Invalid RPN expression")
	return stack[0]

if __name__ == '__main__':
	print(evalRPN(["2", "1", "+", "3", "*"]))  # Output: 9
	print(evalRPN(["4", "13", "5", "/", "+"]))  # Output: 6
	print(evalRPN(["10", "6", "9", "3", "/", "-", "*"]))  # Output: 70
	print(evalRPN(["2", "1", "+"]))  # Output: 3
	print(evalRPN(["4", "13", "5", "/", "+"]))  # Output: 6

File: java-way/src/main/java/com/crack/snap/make/medium/stack/EvaluateReversePolishNotation.java
package com.crack.snap.make.medium.stack;

import java.util.List;
import java.util.Stack;

/**
 * @author Mohan Sharma
 */
public class EvaluateReversePolishNotation {

    public int evalRPN(String[] tokens) {
        var stack = new Stack<Integer>();
        for (var token: tokens) {
            if (isOperator(token)) {
                var b = stack.pop();
                var a = stack.pop();
                stack.push(calculate(a, b, token));
            } else {
                stack.push(Integer.parseInt(token));
            }
        }
        return stack.pop();
    }

    private boolean isOperator(final String token) {
        var operators = List.of("+", "-", "*", "/");
        return operators.contains(token);
    }

    private int calculate(final Integer a, final Integer b, final String operator) {
        return switch (operator) {
            case "+" -> a + b;
            case "-" -> a - b;
            case "*" -> a * b;
            case "/" -> a / b;
            default -> throw new IllegalStateException("Unexpected operator: " + operator);
        };
    }

    public static void main(String[] args) {
        var tokens = new String[]{"2", "1", "+", "3", "*"};
        var evaluator = new EvaluateReversePolishNotation();
        System.out.println(evaluator.evalRPN(tokens)); // Output: 9
    }
}
